### PR TITLE
REF: implement _apply_blockwise

### PR DESCRIPTION
**Pull request by jbrockmendel**
Preliminary pushing towards #34714

### CODE REVIEW

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -487,6 +487,36 @@ def _get_window_indexer(self, window: int) -> BaseIndexer:
             return VariableWindowIndexer(index_array=self._on.asi8, window_size=window)
         return FixedWindowIndexer(window_size=window)
 
+    def _apply_blockwise(self, homogeneous_func: Callable):
```
Comment (jreback) [inline pandas/core/window/rolling.py:490]: can you type the output

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -487,6 +487,36 @@ def _get_window_indexer(self, window: int) -> BaseIndexer:
...
+        """
+        Apply the given function to the DataFrame broken down into homogeneous
+        sub-frames.
+        """
+        # This isn't quite blockwise, since `blocks` is actually a collection
+        #  of homogenenous DataFrames.
+        blocks, obj = self._create_blocks(self._selected_obj)
+
+        skipped = []
```
Comment (jreback) [inline pandas/core/window/rolling.py:499]: can you type these

**Review by jreback — CHANGES_REQUESTED** [review]
I think ultimately you want to put this _apply_blockwise in a BlockwiseMixin, similar to SelectionMixin, so you can add it to groupby and so on, but suppose could be done later

**Comment by jbrockmendel** [comment]
> I think ultimately you want to put this _apply_blockwise in a BlockwiseMixin, similar to SelectionMixin, so you can add it to groupby and so on, but suppose could be done later

I'm planning a BlockManager method that subsumes all of the apply-blockwise usages.

**Comment by jbrockmendel** [comment]
@simonjayhawkins is this new/expected?

```
pandas/core/window/rolling.py:500: error: Type variable "pandas._typing.ArrayLike" is unbound
pandas/core/window/rolling.py:500: note: (Hint: Use "Generic[ArrayLike]" or "Protocol[ArrayLike]" base class to bind "ArrayLike" inside a class)
pandas/core/window/rolling.py:500: note: (Hint: Use "ArrayLike" in function signature to bind "ArrayLike" inside a function)
```

File: pandas/core/window/rolling.py [hunk 3]
```diff
@@ -487,6 +487,36 @@ def _get_window_indexer(self, window: int) -> BaseIndexer:
             return VariableWindowIndexer(index_array=self._on.asi8, window_size=window)
         return FixedWindowIndexer(window_size=window)
 
+    def _apply_blockwise(self, homogeneous_func: Callable) -> FrameOrSeries:
```
Comment (simonjayhawkins) [inline pandas/core/window/rolling.py:490]:
```suggestion
    def _apply_blockwise(
        self, homogeneous_func: Callable[..., ArrayLike]
    ) -> FrameOrSeries:
```

will get mypy to green. (but not yet checked that the type is correct, just consistent with the variable annotation declaration in the method body)

**Comment by jbrockmendel** [comment]
that worked, thanks @simonjayhawkins

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**